### DISCUSSION TITLE
Consider introducing the concept of initial configuration presets

### DISCUSSION THREAD

**Discussion by kashike** [PaperMC/Paper#10484] [Ideas]
CraftBukkit, Spigot, and Paper have all, for a long time, made changes to the server that changes the behavior to differ from the vanilla game. This is not always wanted by some server owners (see https://github.com/PaperMC/docs/issues/21 and https://github.com/PaperMC/docs/pull/326).

One solution would be to introduce the concept of multiple different configuration presets:
- the "recommended" preset, which would be what our current configuration becomes
- the "vanilla behavior" preset, which sets certain configuration options to the value that restores the behavior to match vanilla

The initial thought for this would be a command line option, such as:
`-Dpaper.configurationPreset=<preset name>`: Sets the configuration preset to be used. If the configuration files have not been generated, the preset specified will be used when generating them. If the configuration does exist, nothing will change.

`-Dpaper.configurationPresetForce=true`: Option to force modifying an existing configuration to match the specified preset.

The [start script generator](https://docs.papermc.io/misc/tools/start-script-gen) could potentially also have an option to specify this option as well.

**Comment by EterNityCH** [comment]
Incorporating it into our existing start script generator sounds like a really good idea. This allows the majority of the server (which are usually less than 10 player concurrently) to enjoy a default that is closer to the Vanilla gameplay experience. 💯

**Comment by BNTFryingPan** [comment]
im sure if this isnt added someone will want it, but a "best performance" preset that optimizes things at the cost of breaking vanilla features would be good. of course some settings exist that might need to be fine-tuned by a server owner based on hardware or player behavior, but these settings could be excluded.

i also just wanted to throw this out there, even if it might be a bit out of scope, but if its possible to create presets your own presets (which im sure someone will have some reason to), perhaps they could work like resource packs. a preset can set a value for a configuration option, or could not include it at all, and it fallback to the default. this could allow for complex setups that combine presets in multiple ways, but may also be a bit out of scope.

**Comment by electronicboy** [comment]
> im sure if this isnt added someone will want it, but a "best performance" preset that optimizes things at the cost of breaking vanilla features would be good. 

This just reminds me of the various optimisation guides over the years which don't really document what a recommended setting they suggest does, or some server owner followed it blindly and now some mechanism doesn't work because they explicitly disabled the very thing they're testing, etc...

**Comment by Lucavon** [comment]
The proposal suggests that nothing should happen if the config already exists and the argument is added. This, in my opinion, can be confusing in two ways:

1) The player adds it to an existing server, expecting the e.g. vanilla config to take effect, but it doesn't. 
2) The player has the argument when initially setting up the server, but then removes it, expecting Paper to use the default configuration, while it keeps the pregenerated profile-based configuratuon.

While I don't have any solutions for these issues, I wanted to make sure everyone's aware of these potential pitfalls.

**Comment by notTamion** [comment]
I think when the properties are properly documented these are only going to be issues if you didn't read the documentation about them at all. Maybe this could be mitigated by changing the naming to something like -Dpaper.intialConfigurationPreset or printing a warning if the server is started with the property while configurations already exist.

**Comment by Lucavon** [comment]
That seems like a much better option, because most likely, the average user will just find what they are looking for on Google and copy paste it without reading the docs. A name like that would make it clearer it's only for the initial configuration even to those who didn't read up on it.

**Comment by EterNityCH** [comment]
I believe that we will be able to assist the concerns on our support channel. We have an awesome and active supoort community to take care of this type of situation.
